String.Split() trim last substring if specified

lordfanger opened the pull request:
When String.Split() is called with null/empty array and whitespaces are used as separators, count is specified and TrimEntries are set the result is unexpected.

Actual behavior:
var x = "a b ".Split(Array.Empty<char>(), 2, StringSplitOptions.TrimEntries); // x is ["a", "b "]
The last substring is not trimmed from end.

Changed behavior:
var x = "a b ".Split(Array.Empty<char>(), 2, StringSplitOptions.TrimEntries); // x is ["a", "b"]
The last substring is trimmed correctly.

lordfanger committed:
String.Split() trim last substring if specified

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=81331) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=81331) before we can accept your contribution.<br/><br/>:x: lordfanger [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=81331)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=81331) it.</sub>

danmoseley commented:
Is this long standing behavior, e.g. .NET Framework even? Could code be broken by this change?

am11 commented:
> Is this long standing behavior, e.g. .NET Framework even? Could code be broken by this change?

The behavior is present since`TrimEntries` field was added to `StringSplitOptions` in .NET 5.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:
```diff
@@ -1507,8 +1507,8 @@ private string[] SplitInternal(ReadOnlySpan<char> separators, int count, StringS
             }
-
-            if (separators.IsEmpty)
+            // The TrimEntries flag must be left intact if we are constrained by count because we need to process last substring.
+            if (separators.IsEmpty && count > Length)
             {
                 // Caller is already splitting on whitespace; no need for separate trim step
                 options &= ~StringSplitOptions.TrimEntries;
```
```suggestion

            if (separators.IsEmpty && count > Length)
            {
                // Caller is already splitting on whitespace; no need for separate trim step if the count is sufficient
                // to examine the whole input.
                options &= ~StringSplitOptions.TrimEntries;
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:
```diff
@@ -3354,7 +3354,8 @@ public static int SplitAny(this ReadOnlySpan<char> source, Span<Range> destinati
 
             // If the separators list is empty, whitespace is used as separators.  In that case, we want to ignore TrimEntries if specified,
             // since TrimEntries also impacts whitespace.
-            if (separators.IsEmpty)
+            // The TrimEntries flag must be left intact if we are constrained by count because we need to process last substring.
```
Nit: here and below, please remove the newline between the existing sentence and the new one

lordfanger replied:
Will do.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:
```diff
@@ -3354,7 +3354,8 @@ public static int SplitAny(this ReadOnlySpan<char> source, Span<Range> destinati
             // since TrimEntries also impacts whitespace.
-            if (separators.IsEmpty)
+            // The TrimEntries flag must be left intact if we are constrained by count because we need to process last substring.
+            if (separators.IsEmpty && destination.Length > source.Length)
```
Why > rather than >= ?  Is there a corner-case that fails if >= is used?

lordfanger replied:
I used > because in tests was `new Range[value.Length + 1]` for count == int.MaxValue.

I did some testing and >= is fine. `destination.Length >= source.Length - 1` is first to break corner-case I could think of.

lordfanger replied:
Should I replace > with >=?

stephentoub replied:
You can leave it as >.  It's already likely a large overestimate, and it's easier to explain the reasoning that you can't possibly have N+1 segments for N characters.

lordfanger committed:
Update src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

lordfanger committed:
doc update

lordfanger commented:
> Is this long standing behavior, e.g. .NET Framework even? Could code be broken by this change?

It could break code that would rely on that behavior but it'll be very rare. But it needs all 3 conditions to be true:

- Split by whitespace (rare)
- Limit by count (not so often)
- TrimEntries (only 2 years old)

lordfanger committed:
Merge branch 'dotnet:main' into string-split

dakersnar commented:
This does seem pretty unintuitive.

I was poking around at this functionality and discovered that if you limit the count to 1, it correctly trims the whitespace. Any idea why that case works but 2 (or higher) doesn't?

var x = "a b ".Split(Array.Empty(), 1, StringSplitOptions.TrimEntries); // x is ["a b"]

dakersnar reviewed (approved):
LGTM if we are ok with the breaking change.

danmoseley commented:
too obscure to deserve breaking changes label / doc perhaps?

lordfanger commented:
I don't expect a lot broken codes (if any). The workaround is very simple. Trim the input string and it is ok (for performance friendly MemoryExtension overload it is super cheap anyway). No need to fill github issue perhaps. I was notified by failing tests for my IEnumerable<stirng> version of string.Split() after upgrade to .NET 6 where compared results against core. Some cases were already fixed in #73194 but after upgrade to .NET 7 there were still some failing tests. 

The doc already say that the substring will be trimmed if option flag is set and it is what this bugfix does. 

I've looked at docs once more and only updates I suggests are for string.Split(char, int, StringSplitOptions) and string.Split(string, int, StringSplitOptions) where last statement in remarks says "If the string has already been split `count` - 1 times, but the end of the string has not been reached, then the last string in the returned array will contain this instance's remaining trailing substring, untouched.". But the last substring is touched if ends with whitespace and TrimEntries flag is set (these methods are untouched by this PR).

lordfanger commented:
> I was poking around at this functionality and discovered that if you limit the count to 1, it correctly trims the whitespace. Any idea why that case works but 2 (or higher) doesn't?

The count 1 is special case and separator is ignored there, only postprocessing is performed if any (possible trimming is not mentioned in docs).

stephentoub commented:
The previous behavior was a bug; the method clearly wasn't doing what it was supposed to be doing (the caller asked for all of the segments to be trimmed and one wasn't in some corner cases). We don't need to proactively doc it.
